python/ray/data/datatype.py

```diff
@@ -0,0 +1,217 @@
+
+        Returns:
+            DataType: A DataType wrapping the given PyArrow type
+
+        Examples:
+            >>> import pyarrow as pa
+            >>> from ray.data.datatype import DataType
+            >>> DataType.from_arrow(pa.timestamp('s'))
+            DataType(arrow:timestamp[s])
+            >>> DataType.from_arrow(pa.int64())
+            DataType(arrow:int64)
+        """
+        return cls(_internal_type=arrow_type)
+
+    @classmethod
+    def from_numpy(cls, numpy_dtype: Union[np.dtype, str]) -> "DataType":
+        """Create a DataType from a NumPy dtype.
+
+        Args:
+            numpy_dtype: A NumPy dtype object or string representation
+
+        Returns:
+            DataType: A DataType wrapping the given NumPy dtype
+
+        Examples:
+            >>> import numpy as np
+            >>> from ray.data.datatype import DataType
+            >>> DataType.from_numpy(np.dtype('int32'))
+            DataType(numpy:int32)
+            >>> DataType.from_numpy('float64')
+            DataType(numpy:float64)
+        """
+        if isinstance(numpy_dtype, str):
+            numpy_dtype = np.dtype(numpy_dtype)
+        return cls(_internal_type=numpy_dtype)
+
+    @classmethod
+    def from_python(cls, python_type: type) -> "DataType":
+        """Create a DataType from a Python type.
+
+        Args:
+            python_type: A Python type object
+
+        Returns:
+            DataType: A DataType wrapping the given Python type
+
+        Examples:
+            >>> from ray.data.datatype import DataType
+            >>> DataType.from_python(int)
+            DataType(python:int)
+            >>> DataType.from_python(str)
+            DataType(python:str)
+        """
+        return cls(_internal_type=python_type)
+
+    @classmethod
+    def infer_dtype(cls, value: Any) -> "DataType":
+        """Infer DataType from a Python value, handling numpy, Arrow, and Python types.
+
+        Args:
+            value: Any Python value to infer the type from
+
+        Returns:
+            DataType: The inferred data type
+
+        Examples:
+            >>> import numpy as np
+            >>> from ray.data.datatype import DataType
+            >>> DataType.infer_dtype(5)
+            DataType(arrow:int64)
+            >>> DataType.infer_dtype("hello")
+            DataType(arrow:string)
+            >>> DataType.infer_dtype(np.int32(42))
+            DataType(numpy:int32)
+        """
+        # 1. Handle numpy arrays and scalars
+        if isinstance(value, (np.ndarray, np.generic)):
+            return cls.from_numpy(value.dtype)
+        # 2. Handle PyArrow scalars
+        elif (
+            hasattr(value, "type")
+            and hasattr(pa, "Scalar")
+            and isinstance(value, pa.Scalar)
+        ):
+            return cls.from_arrow(value.type)
+
+        # 3. Try PyArrow type inference for regular Python values
+        else:
+            try:
+                inferred_arrow_type = pa.infer_type([value])
```

Comment:
Let's use our own inference wrapper handling some of the gotchas: `_infer_pyarrow_type`